Add document for testing component accessibility (#2356)
alphagov/govuk-frontend: docs/contributing/test-components-using-accessibility-acceptance-criteria.md

# Test components using accessibility acceptance criteria

We're still developing this page. To suggest changes, comment on the [GitHub issue for improving this page's content](https://github.com/alphagov/govuk-design-system/issues/1850). Alternatively, you can [contact us directly](https://design-system.service.gov.uk/#support).

This page will help you if you are:

- contributing a new component to the [GOV.UK Design System](https://design-system.service.gov.uk/)
- iterating an existing component

Your component or iteration should meet: 

> - [the GOV.UK Design System’s accessibility acceptance criteria](link-to-Testing-components-against-the-GOV.UK-Design-System’s-accessibility-acceptance-criteria)

Reviewer: This link needs to be updated

Author: My bad! Thanks for flagging.
```suggestion
- [the GOV.UK Design System’s accessibility acceptance criteria](https://github.com/alphagov/govuk-frontend/blob/071ad5032e1d94f6fbcf34fd248c062c923b792d/docs/contributing/test-components-using-accessibility-acceptance-criteria.md#govuk-design-system---accessibility-acceptance-criteria)
```

> - [the GOV.UK Design System’s accessibility acceptance criteria](https://github.com/alphagov/govuk-frontend/blob/071ad5032e1d94f6fbcf34fd248c062c923b792d/docs/contributing/test-components-using-accessibility-acceptance-criteria.md#govuk-design-system---accessibility-acceptance-criteria)

Reviewer: @EoinShaughnessy Because this is a link to a heading within the same document, we should be able to do something like this:

```suggestion
- [the GOV.UK Design System’s accessibility acceptance criteria](#govuk-design-system---accessibility-acceptance-criteria)
```

Author: Ah great, thanks @vanitabarrett

> - custom accessibility acceptance criteria that you need to write specifically for your component or iteration(#writing-custom-accessibility-acceptance-criteria-for-your-component)

Author: ```suggestion
- [custom accessibility acceptance criteria that you need to write specifically for your component or iteration](#writing-custom-accessibility-acceptance-criteria-for-your-component)
```

> - accessibility acceptance criteria that you need to write specifically for your component or iteration(link-to-Writing-accessibility-acceptance-criteria-for-your-component)

Reviewer: Looks like this is intended to be a link too? Link text needs defining and the link URL updating

Author: ```suggestion
- custom accessibility acceptance criteria that you need to write specifically for your component or iteration(https://github.com/alphagov/govuk-frontend/blob/071ad5032e1d94f6fbcf34fd248c062c923b792d/docs/contributing/test-components-using-accessibility-acceptance-criteria.md#writing-custom-accessibility-acceptance-criteria-for-your-component
```

> - custom accessibility acceptance criteria that you need to write specifically for your component or iteration(https://github.com/alphagov/govuk-frontend/blob/071ad5032e1d94f6fbcf34fd248c062c923b792d/docs/contributing/test-components-using-accessibility-acceptance-criteria.md#writing-custom-accessibility-acceptance-criteria-for-your-component

Reviewer: @EoinShaughnessy Same comment as above, as this is a link to a heading in the same document we can use an [anchor link](https://gist.github.com/asabaylus/3071099), i.e: you can just have everything after the 'hashtag' #, you don't need the full URL.

You're also missing the markdown link syntax here, e.g: [custom accessibility acceptance...etc](link path)

Author: ```suggestion
- custom accessibility acceptance criteria that you need to write specifically for your component or iteration(#writing-custom-accessibility-acceptance-criteria-for-your-component
```

## GOV.UK Design System - accessibility acceptance criteria

Before we publish a component on the GOV.UK Design System, it should meet the following accessibility requirements.

### General accessibility acceptance criteria

To show clearly to users, components should:

- have text with a colour contrast of at least 4.5:1 between text and the background
- not depend on colour alone to communicate information
- not move (for example, flash or blink)
- be visually distinct from other elements on the page
- use [semantic markup](https://www.w3.org/WAI/WCAG21/Techniques/general/G115.html)
- if necessary, hide and display content consistently for both screen reader users and sighted users - [learn more about hiding and displaying content](https://webaim.org/techniques/css/invisiblecontent/#intro)

To make sure users can use them, components should at least be operable with:

- a keyboard
- voice
- touch
- a mouse
- the rotor or equivalent, if your screen reader has one - [learn how to use the rotor and local and global context menus](https://dequeuniversity.com/screenreaders/)

> a screen reader in browse or reader mode

Reviewer: ```suggestion
- a screen reader in browse or reader mode
```

> - any other modes your screen reader might have - [learn how to use different modes](https://dequeuniversity.com/screenreaders/)

Reviewer: Just wanted to check - are we intentionally linking to https://dequeuniversity.com/screenreaders/ twice?

Author: Yeah, that's intentional. Bit clunky, I know. Our rationale was, including it twice hopefully makes it less likely users will pass over accessing content that could help them.

Behaviour may vary between different screen reader modes, which could affect what meets the definition of a component being 'operable'.

To give users the information they need, components should:

- contain markup in an order that makes sense to users
- announce any changes to all users - [see our accessibility workshop slides for an example about informing users](https://www.google.com/url?q=https://docs.google.com/presentation/d/1tPocE9yr6VyE33Y_xzeV4jsHbU6z3vla3IJT1Gqwyoc/edit%23slide%3Did.ge5d201f83b_0_6&sa=D&source=editors&ust=1629108717555636&usg=AOvVaw1nrYs6mVY6yelTob5NRCzC)

> - be labelled as a [landmark region](https://www.w3.org/TR/wai-aria-practices/examples/landmarks/main.html), if appropriate

Reviewer: This page seems to be talking specifically about the `main` landmark, but there are other landmark regions. Can we link anywhere more generic? Maybe https://www.w3.org/TR/wai-aria-practices/examples/landmarks/index.html ?

Author: ```suggestion
- be labelled as a [landmark region](https://www.w3.org/TR/wai-aria-practices/examples/landmarks/index.html), if appropriate
```
